src/System.Collections.Immutable/tests/ImmutableInterlockedTests.cs

```diff
@@ -169,22 +169,47 @@ public void InterlockedExchangeArrayNonDefault()
         }
 
         [Fact]
-        public void InterlockedCompareExchangeArrayDefault()
+        public void InterlockedCompareExchangeArray_LocationDefault_EqualToComparand()
         {
             ImmutableArray<int> array = default(ImmutableArray<int>);
-            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), default(ImmutableArray<int>));
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(default(ImmutableArray<int>), oldValue);
+            Assert.Equal(new[] { 4, 5, 6 }, array);
+        }
+
+        [Fact]
+        public void InterlockedCompareExchangeArray_LocationDefault_NotEqualToComarand()
+        {
+            ImmutableArray<int> array = default(ImmutableArray<int>);
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), ImmutableArray.Create<int>(1, 2, 3));
+            Assert.Equal(default(ImmutableArray<int>), oldValue);
+            Assert.Equal(default(ImmutableArray<int>), array);
         }
 
         [Fact]
-        public void InterlockedCompareExchangeArray()
+        public void InterlockedCompareExchangeArray_LocationNotDefault_EqualToComparand()
+        {
+            ImmutableArray<int> array = ImmutableArray.Create<int>(1, 2, 3);
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(new[] { 1, 2, 3 }, oldValue);
+            Assert.Equal(new[] { 4, 5, 6 }, array);
+        }
+
+        [Fact]
+        public void InterlockedCompareExchangeArray_LocationNotDefault_NotEqualToDefaultComparand()
         {
             ImmutableArray<int> array = ImmutableArray.Create(1, 2, 3);
             var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), default(ImmutableArray<int>));
-            Assert.Equal(array.AsEnumerable(), oldValue);
+            Assert.Equal(new[] { 1, 2, 3 }, oldValue);
+            Assert.Equal(new[] { 1, 2, 3 }, array);
+        }
 
-            var arrayBefore = array;
-            oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
-            Assert.Equal(oldValue.AsEnumerable(), arrayBefore);
+        [Fact]
+        public void InterlockedCompareExchangeArray_LocationNotDefault_NotEqualToNonDefaultComparand()
+        {
+            ImmutableArray<int> array = ImmutableArray.Create(1, 2, 3);
+            var oldValue = ImmutableInterlocked.InterlockedCompareExchange(ref array, ImmutableArray.Create<int>(4, 5, 6), array);
+            Assert.Equal(new[] { 1, 2, 3 }, oldValue);
```

Comment:
Is there already a test that verifies the behavior when the array references are different but contain the same values?